core/src/main/scala/sttp/tapir/typelevel/RuntimeMatch.scala

```diff
@@ -0,0 +1,46 @@
+package sttp.tapir.typelevel
+
+import magnolia.{CaseClass, Magnolia, SealedTrait}
+
+import scala.reflect.ClassTag
+
+trait MatchType[T] {
+  def apply(a: Any): Boolean
+}
+
+trait GenericMatchType {
+  type Typeclass[T] = MatchType[T]
+
+  def combine[T: ClassTag](ctx: CaseClass[Typeclass, T]): Typeclass[T] = { value: Any =>
+    val ct = implicitly[ClassTag[T]]
+    ct.runtimeClass.isInstance(value) && ctx.parameters.forall {
+      param => param.typeclass(param.dereference(value.asInstanceOf[T]))
```

Comment:
why do you need the casts here? isn't `param.typeclass` the same type as `param.dereference(value)`?

Reply:
The think is param.typclass is not directly a MatcherType but a `Typecalss[SType]` and cast is not a real cast because sub.cast return a `SType`.

In other word, `param.typeclass`, `param.typeclass` are not really what they look like. 
Honestly, I discover Magnolia today, and Scala is a bit new too me, so I am definitely not a expert.

But my intention is that param.typeclass need a typed value in order to detemine which typeclass to use.